File: src/test/java/net/scriptgate/masonry/demo/ColorBrick.java
package net.scriptgate.masonry.demo;

import net.scriptgate.common.Color3f;
import net.scriptgate.engine.Renderer;
import net.scriptgate.masonry.api.Brick;
import net.scriptgate.masonry.transition.TransitionComponent;

import static net.scriptgate.common.Color3f.BLACK;
import static net.scriptgate.masonry.transition.LinearTransition.LINEAR;

public class ColorBrick implements Brick {

    private int height;
    private int width;

    private Color3f color;

    private TransitionComponent transitionComponent = new TransitionComponent(LINEAR);


    public ColorBrick(int width, int height) {
        this.width = width;
        this.height = height;
        this.color = new Color3f(1, 0, 0);
    }

    @Override
    public void goTo(int x, int y) {
        transitionComponent.goTo(x, y);
    }

    @Override
    public int getHeight() {
        return height;
    }

    @Override
    public int getWidth() {
        return width;
    }

    @Override
    public boolean isLayoutInstant() {
        return transitionComponent.isLayoutInstant();
    }

    @Override
    public int getX() {
        return transitionComponent.getX();
    }

    @Override
    public int getY() {
        return transitionComponent.getY();
    }

    public Color3f getColor() {
        return color;
    }

    @Override
    public void moveTo(int x, int y) {
        transitionComponent.moveTo(x, y);
    }

    public void update(double elapsedTime) {
        transitionComponent.update(elapsedTime);
    }

    public void render(Renderer renderer) {
        int strokeWidth = 2;

        renderer.setColor(BLACK);
        renderer.fillRect(getX(), getY(), getWidth(), getHeight());
        renderer.setColor(getColor());
        renderer.fillRect(
                getX() + strokeWidth,
                getY() + strokeWidth,
                getWidth() - 2 * strokeWidth,
                getHeight() - 2 * strokeWidth);
    }
}

File: src/test/kotlin/net/scriptgate/masonry/demo/ColorBrick.kt
package net.scriptgate.masonry.demo

import net.scriptgate.common.Color3f
import net.scriptgate.common.Color3f.BLACK
import net.scriptgate.engine.Renderer
import net.scriptgate.masonry.api.Brick
import net.scriptgate.masonry.transition.LinearTransition
import net.scriptgate.masonry.transition.TransitionComponent

class ColorBrick(override val width: Int, override val height: Int) : Brick {

    val color: Color3f = Color3f(1f, 0f, 0f)

    private val transitionComponent = TransitionComponent(LinearTransition.LINEAR)

    override val isLayoutInstant: Boolean
        get() = transitionComponent.isLayoutInstant

    override val x: Int
        get() = transitionComponent.x

    override val y: Int
        get() = transitionComponent.y


    override fun goTo(x: Int, y: Int) {
        transitionComponent.goTo(x, y)
    }

    override fun moveTo(x: Int, y: Int) {
        transitionComponent.moveTo(x, y)
    }

    fun update(elapsedTime: Double) {
        transitionComponent.update(elapsedTime)
    }

    fun render(renderer: Renderer) {
        val strokeWidth = 2

        renderer.setColor(BLACK)
        renderer.fillRect(x, y, width, height)
        renderer.setColor(color)
        renderer.fillRect(
                x + strokeWidth,
                y + strokeWidth,
                width - 2 * strokeWidth,
                height - 2 * strokeWidth)
    }
}
